# Atria

An ultra-fast and accurate adapter and quality trimming software designed for paired-end sequencing data.

Author: Jiacheng (Eric) Chuan

## Usage

Try `atria -h` or `atria --help` for more information.

### Input and Output

The input files should be paired-end FastQ(.gz) files (in the same order):

1. Read 1 files: `-r X_R1.FQ Y_R1.FQ.GZ ...`

2. Read 2 files: `-R X_R2.FQ Y_R2.FQ.GZ ...`

Output all files to a directory: `-o PATH` or `--output-dir PATH`. Default is the current directory.

### Trimming methods

Atria integrated several trimming and read filtration methods. It does the following sequentially.

1. **Adapter Trimming**

   - specify read 1 adapter: `-a SEQ` or ` --adapter1 SEQ` (default: AGATCGGAAGAGCACACGTCTGAACTCCAGTCA)

   - specify read 2 adapter: `-A SEQ` or ` --adapter2 SEQ` (default: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT)

   - disable: `--no-adapter-trim`

2. **Paired-end Consensus Calling**: the overlapped regions of read pairs are checked and corrected. *It is available only when Adapter Trimming is on.*
   - disable: `--no-consensus`

3. **Hard Clip 3' end**: resize reads to a fixed length by discarding extra bases in 3' end.
   - specify the number of bases to keep: `-C INT` or `--clip-after INT` (default: disabled)

4. **Hard Clip 5' end**: remove the first INT bases from 5' end.
   - specify the number of bases to remove: `-c INT` or `--clip5 INT` (default: disabled)

5. **Quality Trimming**: trim low-quality tails. (Trimming read tails when the average quality of bases in a sliding window is low.)

   - specify average quality threshold: `-q 20` or `--quality-score 20` (default: 20)

   - specify sliding window length: `--quality-kmer 5` (default: 5)

   - specify FastQ quality format: `--quality-format Illumina1.8`, or `--quality-format 33` (default: 33, ie. Illumina1.8)

   - disable: `--no-quality-trim`

6. **Tail N Trimming**: trim N tails.
   - disable: `--no-tail-n-trim`

7. **N Filtration**: discard a read pair if the number of N in one read is greater than a certain amount. N tails are ignored if Tail N Trimming is on.
   - specify # N allowed in each read: `-n 15` or `--max-n 15` (default: 15)

   - disable: `-n -1` or `--max-n -1`

8. **Read Length Filtration**: filter read pair length in a range.
   - specify read length range: `--length-range 50:500` (default: 50:500)

   - disable: `--no-length-filtration`

### Parallel computing

Atria has two parallel methods: **multi-threading** and **multi-core**. You can enable them both.

1. **Multi-threading** uses multiple threads to process one sample. The speed marginal gain drops as the threads increase: `-t 8` or `--threads 8`. (Default: 8)
2. **Multi-core** processes multiple samples at the same time. No marginal gain loses if IO and memory are sufficient: `-p 4` or `--procs 4`. (Default: disabled)
3. If memory is not sufficient, use `--log2-chunk-size INT` where INT is from 23 to 25. Memory usage reduces exponentially as it decreases.

### Examples

Print the help page:

```sh
atria --help
```

Trim `R1.fastq(.gz)` and `R2.fastq(.gz)` with default settings:

```sh
atria --read1 R1.fastq(.gz) --read2 R2.fastq(.gz) --adapter1 SEQ1 --adapter2 SEQ2
```

Trim `R1.fastq(.gz)` and `R2.fastq(.gz)` and save output files to `directory`:

```sh
atria --read1 R1.fastq(.gz) --read2 R2.fastq(.gz) --adapter1 SEQ1 --adapter2 SEQ2 \
	--output-dir directory
```

Trim multiple paired samples in parallel mode (using 8 procs):

```sh
atria --read1 S*R1.fastq.gz --read2 S*R2.fastq.gz --adapter1 SEQ1 --adapter2 SEQ2 \
	--procs 2 --threads 4
```

Only perform adapter trimming, skip other trimming and filtration methods:

```sh
atria --read1 R1.fastq(.gz) --read2 R2.fastq(.gz) --adapter1 SEQ1 --adapter2 SEQ2 \
	--no-tail-n-trim --max-n=-1 --no-quality-trim --no-length-filtration
```

Skip quality trimming:

```sh
atria --read1 R1.fastq(.gz) --read2 R2.fastq(.gz) --adapter1 SEQ1 --adapter2 SEQ2 \
	--no-quality-trim
```

### Argument details

```
usage: atria [-t INT] [-p INT] [--log2-chunk-size INDEX]
             -r R1-FASTQ [R1-FASTQ...] -R R2-FASTQ [R2-FASTQ...]
             [-o PATH] [-g AUTO|YES|NO] [--no-adapter-trim] [-a SEQ]
             [-A SEQ] [-T INT] [-d INT] [-D INT] [-s FLOAT] [-S FLOAT]
             [-l INT] [--stats] [--no-consensus]
             [--overlap-score FLOAT] [--min-ratio-mismatch FLOAT]
             [--prob-diff FLOAT] [-C INT] [-c INT] [--no-quality-trim]
             [-q INT] [--quality-kmer INT] [--quality-format FORMAT]
             [--no-tail-n-trim] [-n INT] [--no-length-filtration]
             [--length-range INT:INT] [--version] [-h]

Atria v1.0.2

optional arguments:
  -t, --threads INT     use INT threads to process one sample
                        (multi-threading parallel). (type: Int64,
                        default: 8)
  -p, --procs INT       process at most INT samples at the same time
                        (multi-core parallel) (default: "1")
  --log2-chunk-size INDEX
                        read at most 2^INDEX bits each time. Suggest
                        to process 200,000 reads each time. Reduce
                        INDEX to lower the memory usage. (type: Int64,
                        default: 26)
  --version             show version information and exit
  -h, --help            show this help message and exit

input/output: input read 1 and read 2 should be in the same order:
  -r, --read1 R1-FASTQ [R1-FASTQ...]
                        input read 1 fastq file(s) (if unix & gzip,
                        require `pigz` in environment)
  -R, --read2 R2-FASTQ [R2-FASTQ...]
                        input read 2 fastq file(s) (paired with
                        R1-FASTQ)
  -o, --output-dir PATH
                        store output files and stats to PATH
                        (default: current working directory)
  -g, --gzip AUTO|YES|NO
                        gzip the output fastq files (if unix & gzip,
                        require `pigz` in environment) (default:
                        "AUTO")

adapter trimming:
  --no-adapter-trim     disable adapter and pair-end trimming
  -a, --adapter1 SEQ    read 1 adapter (default:
                        "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA")
  -A, --adapter2 SEQ    read 2 adapter (default:
                        "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT")
  -T, --kmer-tolerance INT
                        # of mismatch allowed in 16-mers matching
                        (type: Int64, default: 2)
  -d, --pe-adapter-diff INT
                        number of bases allowed when disconcordance
                        found between adapter and pair-end search
                        (type: Int64, default: 0)
  -D, --r1-r2-diff INT  number of bases allowed when the insert sizes
                        of r1 and r2 are different (type: Int64,
                        default: 0)
  -s, --r1-r2-score-diff FLOAT
                        if the score difference between the 16-mers of
                        r1 and r2 is greater than FLOAT, the insert
                        size of the low-score read will be adapted to
                        the high-score read (type: Float64, default:
                        3.0)
  -S, --trim-score FLOAT
                        if the final score of the read pair is greater
                        than FLOAT, the reads will be trimmed. (type:
                        Float64, default: 8.0)
  -l, --tail-length INT
                        if the adapter is in the tail region, and
                        insert size of pe match is smaller than this
                        region, do not trim the read. (type: Int64,
                        default: 8)
  --stats               write stats to description lines of r2 reads.

consensus/merging in adapter trimming:
  --no-consensus        disable generating consensus paired reads. If
                        adapter trimming is disabled, consensus
                        calling is not performed even the flag is not
                        set.
  --overlap-score FLOAT
                        if no adapter was found, scan the tails of the
                        paired reads. Then, if the maximum score of
                        the overlapped 16-mers are less than FLOAT,
                        skip consensus calling for the read pair. If
                        adapters are found, this step is ignored.
                        (type: Float64, default: 0.0)
  --min-ratio-mismatch FLOAT
                        if the ratio of mismatch of the overlapped
                        region is less than FLOAT, skip consensus
                        calling. (type: Float64, default: 0.2)
  --prob-diff FLOAT     when doing consensus calling, if the bases are
                        not complementary, the base with the higher
                        quality probability is selected unless the
                        quality probability difference are less than
                        FLOAT (type: Float64, default: 0.0)

hard clipping: trim a fixed length (after adapter trimming):
  -C, --clip-after INT  hard clip the 3' tails to contain only INT
                        bases. 0 to disable. (type: Int64, default: 0)
  -c, --clip5 INT       remove the first INT bases from 5' end. (type:
                        Int64, default: 0)

quality trimming: trim the tail when the average quality of bases in
a sliding window is low (after hard clipping):
  --no-quality-trim     skip quality trimming
  -q, --quality-score INT
                        threshold of quality score; 0 means turn off
                        quality trimming (type: Int64, default: 20)
  --quality-kmer INT    trim the tail once found the average quality
                        of bases in a sliding window is low (type:
                        Int64, default: 5)
  --quality-format FORMAT
                        the format of the quality score (Illumina1.3,
                        Illumina1.8, Sanger, Illumina1.5, Solexa); or
                        the ASCII number when quality score == 0
                        (default: "33")

N trimming (after quality trimming):
  --no-tail-n-trim      disable removing NNNNN tail.
  -n, --max-n INT       # N allowed in each read; N tails not included
                        if --no-tail-n-trim; INT<0 to disable (type:
                        Int64, default: 15)

length filter (after N trimming):
  --no-length-filtration
                        disable length filtration
  --length-range INT:INT
                        length range of good reads; format is min:max
                        (default: "50:500")

Author: Jiacheng (Eric) Chuan

```
