# Atria Benchmark Toolkit

> The toolkit is not used in Atria manuscript. 

> In Atria manuscript, benchmark scripts can be found at https://github.com/cihga39871/Atria/tree/master/benchmark. Skewer-modified ART was used to simulate data (https://sourceforge.net/projects/skewer/files/Simulator/). Stats were collected from script from https://sourceforge.net/projects/skewer/files/Benchmark/.

## Dependency
1. **[Atria](https://github.com/cihga39871/Atria/releases)**
2. **[R](https://www.r-project.org/)** and R packages argparse, plotly, and ggsci. It is required by `atria statplot`.

> `Rscript` is in PATH.
> Use `install.packages("XXX")` in R session to install a package. (Replace XXX with a package name)

## Notes for benchmark

1. When benchmarking **adapter-trimming** with other adapter trimmers, please use **big** fastq files and add `--no-tail-n-trim --max-n=-1 --no-quality-trim --no-length-filtration` to prevent non-adapter trimming. It is recommended to use GNU `time` command to evaluate time and CPU usage. User + system time is a better indicator than elapsed time. Atria also records `program-initializing-time`, `file-initializing-time`, `read-processing-time`, and `post-processing-time` in `*.atria.log.json`.
2. Different trimming options may influence adapter trimming precision and run time. It is not recommended to change the kmer related arguments.

## Benchmark

Atria integrated benchmark programs. See available programs, run `Atria programs`.

```
Available programs:
    atria       Pair-end trimming software (default)
    simulate    Generate artificial pair-end reads
    randtrim    Randomly trim R1 or R2 at a random position
    readstat    Collect trimming statistics
                    (reads should be generated by `atria simulate`)
    statplot    Plot trimming statistics
                    (`Rscript` in PATH required)
    test        Test Atria program
    p | prog    Show this program list
```

### Data simulation

`atria simulate` is designed to simulate paired-end reads. You can customize adapter sequences of R1 and R2, sequence length, repeat times, as well as different insert sizes and error rates.

Print the help page, using `atria simulate -h`

```
usage: <PROGRAM> [-o PREF] [-x REPEAT] [-a SEQ] [-A SEQ]
                 [-s SEQ-LENGTH]
                 [-i INSERT-SIZE-RANGE [INSERT-SIZE-RANGE...]]
                 [-S SUBSITUTION-RATE [SUBSITUTION-RATE...]]
                 [-I INSERTION-RATE [INSERTION-RATE...]]
                 [-D DELETION-RATE [DELETION-RATE...]] [-h]

optional arguments:
  -h, --help            show this help message and exit

output:
  -o, --prefix PREF     prefix of output fastq files (default:
                        "read_simulation")

simulation:
  -x, --repeat REPEAT   repeat times for each case (type: Int64,
                        default: 30000)
  -a, --adapter1 SEQ    read 1 adapter (default:
                        "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA")
  -A, --adapter2 SEQ    read 2 adapter (default:
                        "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT")
  -s, --seq-length SEQ-LENGTH
                        a given sequence length; simulated sequence
                        length might be 1 base more than the value
                        because of simulated phasing error (type:
                        Int64, default: 251)
  -i, --insert-size-range INSERT-SIZE-RANGE [INSERT-SIZE-RANGE...]
                        range of insert size (type: Int64, default:
                        [220, 222, 224, 226, 228, 230, 232, 234, 236,
                        238, 240, 241, 242, 243, 244, 245, 246, 247,
                        248, 249, 250, 251, 256, 257, 258, 259, 260])
  -S, --subsitution-rate SUBSITUTION-RATE [SUBSITUTION-RATE...]
                        subsitution rate per base. it is random for
                        each base. error type includs mismatch (type:
                        Float64, default: [0.0025, 0.01, 0.02])
  -I, --insertion-rate INSERTION-RATE [INSERTION-RATE...]
                        insertion rate; number of arg should be the
                        same as --subsitution-rate (type: Float64,
                        default: [2.0e-5, 5.0e-5, 0.0001])
  -D, --deletion-rate DELETION-RATE [DELETION-RATE...]
                        deletion rate; number of arg should be the
                        same as --subsitution-rate (type: Float64,
                        default: [2.0e-5, 5.0e-5, 0.0001])
```

For example, create paired-end fastq files with the default setting:

```sh
atria simulate -o read_simulation
# ┌ Info: read simulation: output files
# │   r1 = "read_simulation.R1.fastq"
# └   r2 = "read_simulation.R2.fastq"
# ┌ Info: read simulation: all done
# └   elapsed = 2.0790791511535645
```

#### Format of the first line of each read

The first line of every read in the FastQ file records the read simulation attributes. Here is an example:

> @PeReadSimulator2:69:69 TRUE=220 INSERT_SIZE=220 ERROR_RATE=0.00254 SEQ_LENGTH=251 ERROR_INSERT=1 ERROR_ADAPTER=0 SUB=0.0025 INS=2.0e-5 DEL=2.0e-5

It has seven fields:

| Field                     | Description                                                  |
| ------------------------- | ------------------------------------------------------------ |
| @PeReadSimulator:6268:868 | The unique read ID in the FastQ file.                        |
| TRUE=220                  | The length of the real inserted DNA.                         |
| INSERT_SIZE=220           | The length of the original inserted DNA before error simulation. (Indel might be introduced in error simulation, so the real insert size might vary.) |
| ERROR_RATE=0.00254        | The error rate for base simulation.                          |
| SEQ_LENGTH=251            | The raw sequence length.                                     |
| ERROR_INSERT=1            | The number of errors in the inserted DNA.                    |
| ERROR_ADAPTER=0           | The number of errors in the adapter.                         |
| SUB=0.0025                | The error rate of substitution                               |
| INS=2.0e-5                | The error rate of insert                                     |
| DEL=2.0e-5                | The error rate of deletion                                   |



### (Optional) Random-trim test

In some cases, R1 and R2 are not in the same length because of the poly N tails or other problems. If the sequence lengths of R1 and R2 are different, trimmers might not work as well as a uniform sequence length. Random-trim test is designed to examine the performance of trimmers by randomly trimming one of the paired reads at a random position.

It should be used after `randtrim` and before adapter trimming.

 For example, input read_simulation.R1.fastq and read_simulation.R2.fastq.

```sh
atria randtrim read_simulation.R1.fastq read_simulation.R2.fastq
```

The output files are read_simulation.R1.randtrim.fastq and read_simulation.R2.randtrim.fastq.



### Trimming with different software

Running Atria or other trimmers. Please make sure each folder contains results from a *single* timer.

> Please only allow adapter trimming and disable other trimming methods. For Atria: use `--no-tail-n-trim --max-n=-1 --no-quality-trim --no-length-filtration`

```sh
atria -r read_simulation.R1.fastq -R read_simulation.R2.fastq --no-tail-n-trim --max-n=-1 --no-quality-trim --no-length-filtration -o Atria-result
```



### Collecting reads metrics

`Atria readstat` collects adapter-trimming metrics.

```sh
atria readstat */*fastq
# ┌ Info: read simulation stats: output
# │   detail = "Atria-result/read_simulation.R1.atria.fastq.stat-detail.tsv"
# └   summary = "Atria-result/read_simulation.R1.atria.fastq.stat.tsv"
```

For each FastQ file, a detail stats for every read and a summary stats are saved.

#### Format of detail stats file

The detail stats is a tab-delimited matrix:

| Column name        | Description                                                  |
| ------------------ | ------------------------------------------------------------ |
| seq_id             | The read ID. (Letters after a blank space is removed.)       |
| seq_length         | The raw sequence length.                                     |
| insert_size        | The length of the original inserted DNA before error simulation. (Indel might be introduced in error simulation, so the real insert size might vary.) |
| error_rate         | The error rate for base simulation.                          |
| error_insert       | The number of errors in the insert.                          |
| error_adapter      | The number of errors in the adapter.                         |
| true_length        | The real insert size after error simulation.                 |
| trimmed_length     | The read length after adapter trimming.                      |
| delta_length       | true_length - trimmed_length.                                |
| is_trim_successful | If true_length = trimmed_length, true; otherwise, false.     |

#### Format of summary stats file

The summary stats is a tab-delimited matrix:

| Column name         | Description                                                  |
| ------------------- | ------------------------------------------------------------ |
| seq_length          | The read ID. (Letters after a blank space is removed.)       |
| insert_size         | The length of the original insert DNA before error simulation. (Indel might be introduced in error simulation, so the real insert size might vary.) |
| error_rate          | The error rate for base simulation.                          |
| repeat              | The number of read pairs found with the same seq_length, insert_size and error_rate. |
| precision           | The rate of reads trimmed at the true position.              |
| rate_overtrim       | The rate of reads trimmed all adapter letters and some insert DNA letters. |
| rate_undertrim      | The rate of reads with adapter letters untrimmed.            |
| deviation           | The median trimming deviation of imprecisely-trimmed reads.  |
| deviation_overtrim  | The median trimming deviation of over-trimmed reads.         |
| deviation_undertrim | The median trimming deviation of under-trimmed reads.        |





#### Benchmark plots

`atria statplot` is an interactive plotter for summary stats.

> Please separate R1 and R2 in different plots.

```sh
atria statplot -i */*R1*stat.tsv -t ".R1"
# peReadSimulatorStatsPlot: Output: trimmer_performance.preciseness.R1.html
# peReadSimulatorStatsPlot: Output: trimmer_performance.deviation.R1.html
atria statplot -i */*R2*stat.tsv -t ".R2"
# peReadSimulatorStatsPlot: Output: trimmer_performance.preciseness.R2.html
# peReadSimulatorStatsPlot: Output: trimmer_performance.deviation.R2.html
```

**trimmers.preciseness.html** is the benchmark of preciseness.

It has 3 subplots with the same x-axis: original insert size, meaning the length of the simulated DNA fragment before error simulation.

The y-axis of the top subplot is precision, meaning the rate of reads trimmed at the true position.

The y-axis of the mid subplot is over-trimmed rate.

And the y-axis of the bottom subplot is under-trimmed rate.

Usually, higher under-trimmed rate introduces more noise to the subsequent analysis, so it is important. Besides, higher over-trimmed rate removes more informative data.

At the bottom of the figure, you can choose different error rates. Error rate refers to the chance of mismatch/insert/deletion when simulating a nucleotide. Higher error rate means the reads are more variant and require trimmers to be more clever.

**trimmers.deviation.html** is the median deviation of the falsely-trimmed reads. It shows the median distance between the true position and the trimmed position.
